File: s_coding/2178_미로탐색.cpp
#if 0
//미로탐색- 구조체 + queue 쓸때
//https://www.acmicpc.net/problem/2178

#include <iostream>
#include <queue>
#include <algorithm>
using namespace std;

int N, M;//세로, 가로
int map[100 + 10][100 + 10];
int visited[100 + 10][100 + 10];
int dh[] = { -1,1,0,0 }; //상하좌우
int dw[] = { 0,0,-1,1 };

typedef struct {
	int h;
	int w;
	int t;
}info;

int BFS() {
	
	queue <info> q;
	info cur;

	q.push({ 1, 1 ,1 });
	visited[1][1] = 1;

	while (!q.empty()) {

		cur = q.front();
		q.pop();

	
		if (cur.h == N && cur.w == M)return cur.t;

		for (int i = 0; i < 4; i++) {
			int nh = cur.h + dh[i];
			int nw = cur.w + dw[i];

			if (nh <= 0 || nh > N || nw <= 0 || nw > M)continue;
			if (map[nh][nw] == 0 || visited[nh][nw] == 1)continue;
			visited[nh][nw] = 1;
			q.push({ nh,nw,cur.t + 1 });
			//cout << nh << " " << nw << " " <<cur.t+1<< endl;
		}

	}

}
void InputData() {
	cin >> N >> M;

	for (int i = 1; i <= N; i++) {
		for (int j = 1; j <= M; j++) {
			scanf("%1d", &map[i][j]);
		}
	}
}
int main() {
	int sol;

	freopen("input.txt", "r", stdin);

	InputData();

	sol = BFS();
	cout << sol;
	return 0;
}

#endif

File: s_coding/BOJ/2178_미로 탐색.py
#미로 탐색
#https://www.acmicpc.net/problem/2178
import sys
from collections import deque
sys.stdin = open("input.txt","r")
dh=[-1,1,0,0]
dw=[0,0,-1,1]
n,m=map(int,input().split())
board=[list(map(int,input())) for _ in range(n) ]
dis=[[0]*m for _ in range(n)]
Q=deque()
Q.append((0,0)) #출발점
board[0][0]=0
while Q:
    cur=Q.popleft()
    for i in range(4):
        nh=cur[0]+dh[i]
        nw=cur[1]+dw[i]
        if 0<=nh<n and 0<=nw<m and board[nh][nw]==1:
            board[nh][nw]=0
            dis[nh][nw]=dis[cur[0]][cur[1]]+1
            Q.append((nh,nw))
if dis[n-1][m-1]+1==0:
    print(-1)
else:
    print(dis[n-1][m-1]+1)
for x in dis:
    print(x,end=' ')
    print()
